File: backend/app/api/v1/goals.py
from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi import status as http_status
from typing import Optional
from uuid import UUID
from sqlalchemy.ext.asyncio import AsyncSession

from ...database.session import get_db_session
from ...security.dependencies import get_auth_context, require_supervisor_or_above
from ...security.context import AuthContext
from ...schemas.goal import Goal, GoalDetail, GoalList, GoalCreate, GoalUpdate
from ...schemas.common import PaginationParams, BaseResponse
from ...services.goal_service import GoalService
from ...core.exceptions import NotFoundError, PermissionDeniedError, ConflictError, ValidationError, BadRequestError

router = APIRouter(prefix="/goals", tags=["goals"])

@router.get("/", response_model=GoalList)
async def get_goals(
    pagination: PaginationParams = Depends(),
    period_id: Optional[UUID] = Query(None, alias="periodId", description="Filter by evaluation period ID"),
    user_id: Optional[UUID] = Query(None, alias="userId", description="Filter by user ID (supervisor/admin only)"),
    goal_category: Optional[str] = Query(None, alias="goalCategory", description="Filter by goal category (業績目標, コンピテンシー, コアバリュー)"),
    status: Optional[str] = Query(None, description="Filter by status (draft, pending_approval, approved, rejected)"),
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """Get goals for the current user or filtered goals for supervisors/admins."""
    try:
        service = GoalService(session)
        
        result = await service.get_goals(
            current_user_context=context,
            user_id=user_id,
            period_id=period_id,
            goal_category=goal_category,
            status=status,
            pagination=pagination
        )
        
        return result
        
    except NotFoundError as e:
        raise HTTPException(
            status_code=http_status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except PermissionDeniedError as e:
        raise HTTPException(
            status_code=http_status.HTTP_403_FORBIDDEN,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error fetching goals: {str(e)}"
        )

@router.post("/", response_model=Goal)
async def create_goal(
    goal_create: GoalCreate,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """Create a new goal."""
    try:
        service = GoalService(session)
        
        result = await service.create_goal(goal_create, context)
        
        return result
        
    except ValidationError as e:
        raise HTTPException(
            status_code=http_status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(e)
        )
    except BadRequestError as e:
        raise HTTPException(
            status_code=http_status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except PermissionDeniedError as e:
        raise HTTPException(
            status_code=http_status.HTTP_403_FORBIDDEN,
            detail=str(e)
        )
    except ConflictError as e:
        raise HTTPException(
            status_code=http_status.HTTP_409_CONFLICT,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating goal: {str(e)}"
        )

@router.get("/period/{period_id}", response_model=GoalList)
async def get_goals_by_period(
    period_id: UUID,
    pagination: PaginationParams = Depends(),
    user_id: Optional[UUID] = Query(None, alias="userId", description="Filter by user ID (supervisor/admin only)"),
    goal_category: Optional[str] = Query(None, alias="goalCategory", description="Filter by goal category"),
    status: Optional[str] = Query(None, description="Filter by status"),
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """Get goals by evaluation period with optional filters."""
    try:
        service = GoalService(session)
        
        result = await service.get_goals(
            current_user_context=context,
            user_id=user_id,
            period_id=period_id,
            goal_category=goal_category,
            status=status,
            pagination=pagination
        )
        
        return result
        
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error fetching goals by period: {str(e)}")

@router.get("/{goal_id}", response_model=GoalDetail)
async def get_goal(
    goal_id: UUID,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """Get detailed goal information by ID."""
    try:
        service = GoalService(session)
        result = await service.get_goal_by_id(goal_id, context)
        return result
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error fetching goal: {str(e)}")

@router.put("/{goal_id}", response_model=Goal)
async def update_goal(
    goal_id: UUID,
    goal_update: GoalUpdate,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """Update a goal."""
    try:
        service = GoalService(session)
        result = await service.update_goal(goal_id, goal_update, context)
        return result
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=http_status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    except BadRequestError as e:
        raise HTTPException(status_code=http_status.HTTP_400_BAD_REQUEST, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=http_status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error updating goal: {str(e)}")

@router.delete("/{goal_id}", response_model=BaseResponse)
async def delete_goal(
    goal_id: UUID,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """Delete a goal."""
    try:
        service = GoalService(session)
        success = await service.delete_goal(goal_id, context)
        if success:
            return BaseResponse(message="Goal deleted successfully")
        else:
            raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to delete goal")
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except BadRequestError as e:
        raise HTTPException(status_code=http_status.HTTP_400_BAD_REQUEST, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=http_status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error deleting goal: {str(e)}")

@router.post("/{goal_id}/submit", response_model=Goal)
async def submit_goal(
    goal_id: UUID,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """Submit a goal for approval (goal owner only)."""
    try:
        service = GoalService(session)
        
        # Update goal status to pending_approval
        from ...schemas.goal import GoalUpdate, GoalStatus
        goal_update = GoalUpdate(status=GoalStatus.PENDING_APPROVAL)
        result = await service.update_goal(goal_id, goal_update, context)
        
        return result
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=http_status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    except BadRequestError as e:
        raise HTTPException(status_code=http_status.HTTP_400_BAD_REQUEST, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=http_status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error submitting goal: {str(e)}")

# Optional api endpoints
@router.post("/{goal_id}/approve", response_model=Goal)
async def approve_goal(
    goal_id: UUID,
    context: AuthContext = Depends(require_supervisor_or_above),
    session: AsyncSession = Depends(get_db_session)
):
    """Approve a goal (supervisor/admin only)."""
    try:
        service = GoalService(session)
        result = await service.approve_goal(goal_id, context)
        return result
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=http_status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    except BadRequestError as e:
        raise HTTPException(status_code=http_status.HTTP_400_BAD_REQUEST, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=http_status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error approving goal: {str(e)}")

@router.post("/{goal_id}/reject", response_model=Goal)
async def reject_goal(
    goal_id: UUID,
    rejection_reason: str = Query(..., alias="reason", description="Reason for rejection"),
    context: AuthContext = Depends(require_supervisor_or_above),
    session: AsyncSession = Depends(get_db_session)
):
    """Reject a goal (supervisor/admin only)."""
    try:
        service = GoalService(session)
        result = await service.reject_goal(goal_id, rejection_reason, context)
        return result
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=http_status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    except BadRequestError as e:
        raise HTTPException(status_code=http_status.HTTP_400_BAD_REQUEST, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=http_status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error rejecting goal: {str(e)}")


@router.get("/supervisor/pending", response_model=GoalList)
async def get_pending_approvals(
    pagination: PaginationParams = Depends(),
    period_id: Optional[UUID] = Query(None, alias="periodId", description="Filter by evaluation period ID"),
    context: AuthContext = Depends(require_supervisor_or_above),
    session: AsyncSession = Depends(get_db_session)
):
    """Get goals pending approval (supervisor/admin only)."""
    try:
        service = GoalService(session)
        result = await service.get_pending_approvals(context, period_id, pagination)
        return result
    except NotFoundError as e:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDeniedError as e:
        raise HTTPException(status_code=http_status.HTTP_403_FORBIDDEN, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=http_status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    except BadRequestError as e:
        raise HTTPException(status_code=http_status.HTTP_400_BAD_REQUEST, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=http_status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error fetching pending approvals: {str(e)}")

@router.get("/pending-approval", response_model=GoalList)
async def get_pending_approval_alias(
    pagination: PaginationParams = Depends(),
    period_id: Optional[UUID] = Query(None, alias="periodId", description="Filter by evaluation period ID"),
    context: AuthContext = Depends(require_supervisor_or_above),
    session: AsyncSession = Depends(get_db_session)
):
    """Get goals pending approval (supervisor/admin only) - Alias endpoint."""
    return await get_pending_approvals(pagination, period_id, context, session)

File: frontend/src/api/server-actions/goals.ts
'use server';

import { cache } from 'react';
import { revalidateTag } from 'next/cache';
import { goalsApi } from '../endpoints/goals';
import { CACHE_TAGS } from '../utils/cache';
import type {
  UUID,
  GoalCreateRequest,
  GoalUpdateRequest,
  GoalListResponse,
  GoalResponse,
} from '../types';


/**
 * Server action to get goals with optional filtering, pagination, and caching
 */
async function _getGoalsAction(params?: {
  periodId?: UUID;
  userId?: UUID;
  goalCategory?: string;
  status?: string | string[];
  page?: number;
  limit?: number;
  includeReviews?: boolean;
  includeRejectionHistory?: boolean;
}): Promise<{ success: boolean; data?: GoalListResponse; error?: string }> {
  try {
    const response = await goalsApi.getGoals(params);

    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to fetch goals',
      };
    }

    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Get goals action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while fetching goals',
    };
  }
}

export const getGoalsAction = cache(async (params?: {
  periodId?: UUID;
  userId?: UUID;
  goalCategory?: string;
  status?: string | string[];
  page?: number;
  limit?: number;
  includeReviews?: boolean;
  includeRejectionHistory?: boolean;
}): Promise<{ success: boolean; data?: GoalListResponse; error?: string }> => {
  return _getGoalsAction(params);
});


/**
 * Server action to get a single goal by ID with caching
 */
export const getGoalByIdAction = cache(async (goalId: UUID): Promise<{
  success: boolean;
  data?: GoalResponse;
  error?: string;
}> => {
  try {
    const response = await goalsApi.getGoalById(goalId);

    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to fetch goal',
      };
    }

    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Get goal by ID action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while fetching goal',
    };
  }
});
/**
 * Server action to create a new goal with cache revalidation
 */
export async function createGoalAction(data: GoalCreateRequest): Promise<{
  success: boolean;
  data?: GoalResponse;
  error?: string;
}> {
  try {
    const response = await goalsApi.createGoal(data);
    
    if (!response.success || !response.data) {
      const errorMessage = response.error || response.errorMessage || 'Failed to create goal';
      return {
        success: false,
        error: errorMessage,
      };
    }
    
    // Revalidate goals and related caches after successful creation
    revalidateTag(CACHE_TAGS.GOALS);
    revalidateTag(CACHE_TAGS.SELF_ASSESSMENTS);
    revalidateTag(CACHE_TAGS.SUPERVISOR_REVIEWS);
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Create goal action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while creating goal',
    };
  }
}

/**
 * Server action to update an existing goal with cache revalidation
 */
export async function updateGoalAction(id: UUID, data: GoalUpdateRequest): Promise<{
  success: boolean;
  data?: GoalResponse;
  error?: string;
}> {
  try {
    const response = await goalsApi.updateGoal(id, data);
    
    if (!response.success || !response.data) {
      const errorMessage = response.error || response.errorMessage || 'Failed to update goal';
      return {
        success: false,
        error: errorMessage,
      };
    }
    
    // Revalidate goals and related caches after successful update
    revalidateTag(CACHE_TAGS.GOALS);
    revalidateTag(CACHE_TAGS.SELF_ASSESSMENTS);
    revalidateTag(CACHE_TAGS.SUPERVISOR_REVIEWS);
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Update goal action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while updating goal',
    };
  }
}

/**
 * Server action to delete a goal with cache revalidation
 */
export async function deleteGoalAction(id: UUID): Promise<{
  success: boolean;
  error?: string;
}> {
  try {
    const response = await goalsApi.deleteGoal(id);
    
    if (!response.success) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to delete goal',
      };
    }
    
    // Revalidate goals and related caches after successful deletion
    revalidateTag(CACHE_TAGS.GOALS);
    revalidateTag(CACHE_TAGS.SELF_ASSESSMENTS);
    revalidateTag(CACHE_TAGS.SUPERVISOR_REVIEWS);
    
    return {
      success: true,
    };
  } catch (error) {
    console.error('Delete goal action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while deleting goal',
    };
  }
}

/**
 * Server action to submit a goal with status change and cache revalidation
 */
export async function submitGoalAction(id: UUID, status: 'draft' | 'submitted'): Promise<{
  success: boolean;
  data?: GoalResponse;
  error?: string;
}> {
  try {
    const response = await goalsApi.submitGoal(id, status);
    
    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.error || 'Failed to submit goal',
      };
    }
    
    // Revalidate goals cache after status change
    revalidateTag(CACHE_TAGS.GOALS);
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Submit goal action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while submitting goal',
    };
  }
}

/**
 * Server action to approve a pending goal (supervisor action) with cache revalidation
 */
export async function approveGoalAction(goalId: UUID): Promise<{
  success: boolean;
  data?: GoalResponse;
  error?: string;
}> {
  try {
    const response = await goalsApi.approveGoal(goalId);
    
    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.error || 'Failed to approve goal',
      };
    }
    
    // Revalidate goals cache after approval
    revalidateTag(CACHE_TAGS.GOALS);
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Approve goal action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while approving goal',
    };
  }
}
/**
 * Server action to reject a pending goal with reason (supervisor action) and cache revalidation
 */
export async function rejectGoalAction(goalId: UUID, reason: string): Promise<{
  success: boolean;
  data?: GoalResponse;
  error?: string;
}> {
  try {
    const response = await goalsApi.rejectGoal(goalId, reason);
    
    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.error || 'Failed to reject goal',
      };
    }
    
    // Revalidate goals cache after rejection
    revalidateTag(CACHE_TAGS.GOALS);
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Reject goal action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while rejecting goal',
    };
  }
}

/**
 * Server action to get all goals for admin visualization (admin-only)
 *
 * This action fetches ALL users' goals in the organization.
 * Performance: includeReviews defaults to true for batch optimization
 */
async function _getAdminGoalsAction(params?: {
  periodId?: UUID;
  userId?: UUID;
  departmentId?: UUID;
  goalCategory?: string;
  status?: string | string[];
  page?: number;
  limit?: number;
  includeReviews?: boolean;
  includeRejectionHistory?: boolean;
}): Promise<{ success: boolean; data?: GoalListResponse; error?: string }> {
  try {
    const response = await goalsApi.getAdminGoals(params);

    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to fetch admin goals',
      };
    }

    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Get admin goals action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while fetching admin goals',
    };
  }
}

export const getAdminGoalsAction = cache(_getAdminGoalsAction);
